### PR TITLE
Allow user to edit file information directly in the modal

### PR DESCRIPTION
**Pull request by maxime-rainville**
This PR adds the ability to edit the detail of a file within the file insertion modal.

# Other improvement
* Move `getFormSchema` out of AssetAdmin
* Replace convoluted logic to get breadcrumb data into redux store
* Tweak the header spacing

# Parent issue
* #813

### CODE REVIEW

File: client/src/containers/AssetAdmin/AssetAdmin.js [hunk 1]
```diff
@@ -25,62 +24,7 @@ import configShape from 'lib/configShape';
 import { injectGraphql } from 'lib/Injector';
 import BulkDeleteConfirmation from '../BulkDeleteConfirmation/BulkDeleteConfirmation';
 import * as confirmDeletionActions from 'state/confirmDeletion/ConfirmDeletionActions';
-
-function getFormSchema({ config, viewAction, folderId, fileId, type }) {
```
Comment (maxime-rainville) [inline client/src/containers/AssetAdmin/AssetAdmin.js]: `getFormSchema` is required in other places now. I've moved it to its own file.

File: client/src/containers/AssetAdmin/AssetAdmin.js [hunk 2]
```diff
@@ -25,62 +24,7 @@ import configShape from 'lib/configShape';
...
-}
-
-
-/**
- * Check if either of the two objects differ
- *
- * @param {Object} left
- * @param {Object} right
- */
-function compare(left, right) {
```
Comment (maxime-rainville) [inline client/src/containers/AssetAdmin/AssetAdmin.js]: `compare` was only use by some dodgy willReceiveProps. It's no longer used anymore.

File: client/src/containers/AssetAdmin/AssetAdmin.js [hunk 3]
```diff
@@ -138,63 +72,6 @@ class AssetAdmin extends Component {
     return 0;
   }
 
-  /**
-   * Assign breadcrumbs from selected folder
-   *
-   * @param {Object} props
-   */
-  setBreadcrumbs(props) {
```
Comment (maxime-rainville) [inline client/src/containers/AssetAdmin/AssetAdmin.js]: Most of this logic was moved to a dedicated stateless Asset Breadcrumb component.

File: client/src/containers/AssetAdmin/Breadcrumb.js [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import React from 'react';
+import PropTypes from 'prop-types';
+import i18n from 'i18n';
+import { Component as PlainBreadcrumb } from 'components/Breadcrumb/Breadcrumb';
+import { hasFilters } from 'components/Search/Search';
+
+/**
+ * Display a breadcrumb based on the current folder, bypassing the stateful Breadcrumb component.
+ */
+const Breadcrumb = ({ folder, query, getUrl, onBrowse, onFolderIcon }) => {
```
Comment (maxime-rainville) [inline client/src/containers/AssetAdmin/Breadcrumb.js]: The old approach was taking the information we were getting back from our GraphQL request, building some breadcromb data and calling redux actions to store them in the redux store.

This was creating two source of truth and adding useless complexity. Instead, I'm using the plain Breadcrumb component (as opposed to the one that's wired into the redux store) and passing it the breadcrumb data directly.

File: client/src/containers/AssetAdmin/AssetAdmin.js [hunk 5]
```diff
@@ -351,11 +228,8 @@ class AssetAdmin extends Component {
 
   /**
    * Handler for when the folder icon is clicked (to edit hte folder)
```
Comment (maxime-rainville) [inline client/src/containers/AssetAdmin/AssetAdmin.js]:
```suggestion
   * Handler for when the folder icon is clicked (to edit the folder)
```

**Comment by maxime-rainville** [comment]
Things left to do to get this PR over the line:
* [ ] Add unit test for `getFormSchema`
* [ ] Add unit tests to Editor to cover
  * [ ] handleClose
  * [ ] editorHeader
  * [ ] createFn
* [ ] Add unit tests to EditorHeader
* [ ] Update Modal store to cover new state
* [ ] Write some behat test
* [ ] UX feedback
  * [ ] Make placement an header rather than a tab
  * [ ] Add a warning about editing files globally in the modal

File: client/src/containers/Editor/Editor.js [hunk 6]
```diff
@@ -33,6 +43,27 @@ class Editor extends Component {
     };
   }
 
+  /**
+   * Build the form schema URL to pass to the Form Builder Loader
+   * @returns {string}
+   */
+  getFormSchemaUrl() {
```
Comment (maxime-rainville) [inline client/src/containers/Editor/Editor.js]: This used to be done in the rendering method, but I needed it in two places, so I moved it to it's own method. I've switch to using `qs` and `url` rather than directly manipulating string.

File: client/src/containers/Editor/tests/EditorHeader-story.js [hunk 7]
```diff
@@ -0,0 +1,77 @@
+import React from 'react';
```
Comment (maxime-rainville) [inline client/src/containers/Editor/tests/EditorHeader-story.js]: Not sure we should keep this file. It's helpful to tweak the styling of the Editor Header, but it's kind of clunky.

File: client/src/entwine/TinyMCE_sslink-file.js [hunk 8]
```diff
@@ -63,14 +64,21 @@ jQuery.entwine('ss', ($) => {
    */
   $(`.js-injector-boot #${modalId}`).entwine({
     renderModal(isOpen) {
-      const handleHide = () => this.close();
+      const { dispatch } = Injector.reducer.store;
```
Comment (maxime-rainville) [inline client/src/entwine/TinyMCE_sslink-file.js]: This is a tiny bit of mad science. We're getting the redux store in entwine so we can push some data into to control the modal.

File: client/src/state/modal/ModalActionTypes.js [hunk 9]
```diff
@@ -1,3 +1,7 @@
 export default {
-  DEFINE_IMAGE_SIZE_PRESETS: 'DEFINE_IMAGE_SIZE_PRESETS'
+  DEFINE_IMAGE_SIZE_PRESETS: 'DEFINE_IMAGE_SIZE_PRESETS',
```
Comment (maxime-rainville) [inline client/src/state/modal/ModalActionTypes.js]: I'm starting to think that it doesn't make a lot of sense to have the image presets along with all the form schema stuff together. Maybe we need to move it to a different reducer.

File: client/src/containers/Editor/tests/EditorHeader-story.js (hunk 7 above)
Comment (emteknetnz) [reply client/src/containers/Editor/tests/EditorHeader-story.js]: what are the other options for where it could live?

File: client/src/entwine/TinyMCE_sslink-file.js (hunk 8 above)
Comment (emteknetnz) [reply client/src/entwine/TinyMCE_sslink-file.js]: could you put a comment into the code explaining the mad science :)

File: client/src/state/modal/ModalActionTypes.js (hunk 9 above)
Comment (emteknetnz) [reply client/src/state/modal/ModalActionTypes.js]: So how would file naming work, would you split this file into two, as well as ModalActions.js? i.e.
- ModalFormActionType
- ModalImageSizeActionType
- ModalImageSizeActions
- ModalImageSizeActions
?

File: client/src/containers/AssetAdmin/stateRouter.js [hunk 10]
```diff
@@ -111,6 +112,10 @@ class AssetAdminStateRouter extends Component {
       throw new Error(`Invalid action provided: ${action}`);
     }
 
+    if (this.state.fileId !== fileId) {
+      this.props.actions.resetFormStack();
```
Comment (maxime-rainville) [inline client/src/containers/AssetAdmin/stateRouter.js]: This reset the Editor form in modal

File: client/src/entwine/TinyMCE_sslink-file.js (hunk 8 above)
Comment (maxime-rainville) [reply client/src/entwine/TinyMCE_sslink-file.js]: Explain this redux store mad science

File: client/src/state/modal/ModalReducer.js [hunk 11]
```diff
@@ -1,20 +1,72 @@
...
+ */
+
+/**
+ * Initial base state
+ * @type { ModalState }
  */
 export const initialState = {
-  imageSizePresets: []
+  imageSizePresets: [],
+  formSchemaStack: [],
```
Comment (maxime-rainville) [inline client/src/state/modal/ModalReducer.js]: Explain the stack mad logic.

File: tests/behat/features/insert-an-image.feature [hunk 12]
```diff
@@ -36,6 +34,29 @@ Feature: Insert an image into a page
       # Required to avoid "unsaved changed" browser dialog
       And I press the "Save" button
 
+  @foobar
```
Comment (maxime-rainville) [inline tests/behat/features/insert-an-image.feature]:
```suggestion
```

File: client/src/containers/Editor/Editor.js [hunk 13]
```diff
@@ -197,17 +283,22 @@ class Editor extends Component {
     const campaignTitle = i18n._t('Admin.ADD_TO_CAMPAIGN', 'Add to campaign');
     const Loading = this.props.loadingComponent;
 
-    return (<div className={editorClasses.join(' ')}>
+    // FormBuilderLoader doesn't do anything with the file props, but passing it along forces
+    // React to re-render the EditorHeader component when file gets populated by GraphQL data
+    const file = this.props.file;
```
Comment (maxime-rainville) [inline client/src/containers/Editor/Editor.js]: This fixes an issue with the Insert Link form. Where firing a GraphQL request to fetch the file data and a request for the FormSchema of the Insert link form.

Most of the the time, the GraphQL data comes back first and the selected file data gets pass down to the Editor in time for the `EditorHeader` to be rendered correctly. However on some occasion the FormSchema comes back first and the `EditorHeader` gets rendered without the necessary file data.

Passing `file` to `FormBuilderLoader` will cause a re-render when the GraphQL file data is filled in later on, which will eventually cause a re-render of the `EditorHeader`.

File: client/src/containers/Editor/tests/EditorHeader-story.js (hunk 7 above)
Comment (maxime-rainville) [reply client/src/containers/Editor/tests/EditorHeader-story.js]: Just summarising the chat we had yesterday. The reason why this file is "clunky" is because it wrap the EditorHeader into a lot of fluff HTML that is normally generated by the parent Editor component. It's useful for doing basic testing and styling on the EditorHeader component, but not necessarily something we should keep around long term. 

For example, if we update the Editor component, it won't be reflected here, because the HTML is manually written here.

As a side node, this illustrate a weakness of the Editor component. It mixes state management and presentational concerns which make it difficult to test the state management side of it and to create a story for it in the pattern lib.

The conclusion was to delete this file.

File: client/src/state/modal/ModalActionTypes.js (hunk 9 above)
Comment (maxime-rainville) [reply client/src/state/modal/ModalActionTypes.js]: On second thought, I'm tempted to say keep it like this. Moving it to it's own slice of state is not obviously better and there's an argument to say that all this state data is only relevant when AssetAdmin is displayed in a Modal.

Comment (emteknetnz) [reply client/src/state/modal/ModalActionTypes.js]: OK, so that's a definite keep it this way?  Means I can resolve conversation and we can focus on other bits

File: client/src/containers/Editor/Editor.js (hunk 13 above)
Comment (emteknetnz) [reply client/src/containers/Editor/Editor.js]: it's useful to say why a particular piece of code is done the way it is, though most people in the future won't see github comments - would you be able to summarise and add as a code comment?

Comment (maxime-rainville) [reply client/src/containers/Editor/Editor.js]: That's what the two line above are meant to explain more succinctly. Do you think this is not clear enough?
```
    // FormBuilderLoader doesn't do anything with the file props, but passing it along forces
    // React to re-render the EditorHeader component when file gets populated by GraphQL data
```

File: client/src/entwine/TinyMCE_sslink-file.js (hunk 8 above)
Comment (maxime-rainville) [reply client/src/entwine/TinyMCE_sslink-file.js]: I've added a comment about the redux bit.

File: client/src/state/modal/ModalActionTypes.js (hunk 9 above)
Comment (maxime-rainville) [reply client/src/state/modal/ModalActionTypes.js]: Yeah good to resolve. (FYI Sometime I'll leave some explanation about a decision I've made for which I feel ambivalent.)

**Comment by maxime-rainville** [comment]
GGRGRGRGR!!!!

This seems to have broken the UploadField!!!

**Comment by maxime-rainville** [comment]
Fixed the UploadField
Added some Behat test to cover it
Rebased with `1` branch

**Comment by emteknetnz** [comment]
image

I found the text 'Details' on this button to be pretty confusing.  I'm thinking we should change it to 'Original image' or something similar.  Might be worth checking with PUX

**Comment by emteknetnz** [comment]
This relates to the jonom/focuspoint module which was listed on the original issue

'Details' screen = editing original image
'Placement' screen = editing instance i.e. I just clicked the image in wysiwyg > edit image button

When I set the coordinates in the 'Details' screen, then when I click 'save/publish' it works as expected and the coordinates are saved

Then when I see the focus point in the 'Placement' screen, it now has the same coordinates as in the one in the 'Details' screen.   When I change the coordinates and then click 'update file', the coordinates are then reverted to the ones in the 'Details' screen

Possibly we just need to stop the focus point module from appearing in the 'Placement' screen

**Comment by emteknetnz** [comment]
The word 'Details' is doubly confusing when clicking 'Insert from Files' > 'Upload file' (though to be fair it was probably already like this

When I upload an image I just uploaded, when I hover over it, it says 'Details'

When I click the image, I go to the 'Placement' screen.  There's now also a new button that says 'Details' in the top right.  Which is extremely confusing.  Previously this was a 'Details' tab

**Comment by emteknetnz** [comment]
This is sort of an existing issue, though the new interface makes it seem broken

When I click 'Insert from Files' > 'Upload file', and I upload a new image

In both the new and old interface, I'm in the 'Placement' panel and there's a little orange 'Draft' in the top next to the filesize

In the old interface, the natural thing to do was click 'Insert image' which would take me back to the page edit screen.  When I then 'Publish' the page it's on.  The newly uploaded image would still be 'draft' though it didn't really matter because I could still see it from ?stage=Live

In the new interface, after I upload the image, when I goto to the 'Placement' screen and click the 'Details' button, and I click the 'Publish' button (which was not available on the 'Details' tab of the old interface) then I'm redirected back to the 'Placement' panel, but the orange 'draft' text is still there.

So it seems like it failed to publish.  However, if I refresh my browser, and navigate back to where I was, I can see that the file was actually published.

**Comment by maxime-rainville** [comment]
# Wording of "Details" button
@silverstripeux Do we want to rethink the wording of this word.

# jonom/focuspoint module
Didn't think of that. I'll need to have a look at how that module inject their field into the forms.

Technically, that's not a supported module and we're not changing any APIs in a non backward compatible way ... I wouldn't consider changing the Fields in a Form to be an API change.

It is a very popular module however, so I'll see what I can do to make sure it still works. Arguably, the problem here is that we don't have a sensible way for third parties to modify these File/Image form.

# Sticky Draft tag
I think the problem is that I'm stashing the fields, so I might be accidentally stashing the publication status of the file. I'll have a look.

**Comment by clarkepaul** [comment]
@maxime-rainville Had a quick chat to @emteknetnz about some of the things discussed above. We agreed that the wording on the hover state of files needs to change from "Details" to "View" (along with icon change to an eye) because when placing images it no longer takes you to the details of the file. This change would be visible in both File Admin and in the modal when selecting a file to place.
Pasted_Image_30_01_20__3_03_PM

As for changing the text in the button from "Details" to something else, I'm pretty sure Jared went through user testing around the office to get that wording with different designs. It used to be "Edit original" and I think people who were not used to the CMS didn't really understand the concept of an "original" file. I'm happy to stick with "Details" as the label because it essentially takes you to the Details tab of the file.

**Comment by clarkepaul** [comment]
@maxime-rainville notes for the uploadfield UI, sorry move to a better place if not wanted here.

Uploadfield UX for discussion:
1. Bug - File details (name) isn’t updating when it is edited, saved and inserted into a page.
Pasted_Image_30_01_20__4_40_PM
3. When editing the details of a file... can we hide all of the checkboxes on other files, you only need to select one file... and can we hide/remove the bulk actions when editing a file (bottom left corner) as they aren't needed, all those actions are available on the edit panel for a single file.
Pasted_Image_30_01_20__4_34_PM

I haven't yet seen the multi-file experience into the uploadfield but I don't see that the bulk actions are needed there either, but the checkboxes are required to select multiple files.

With those changes, I think we have progressed things enough to release and provide enough benefit to users.

**Comment by emteknetnz** [comment]
Re Sticky Draft tag - related issue https://github.com/silverstripe/silverstripe-asset-admin/issues/960

**Comment by maxime-rainville** [comment]
I've fixed the following two points:
* Wording of "Details" button
* Sticky Draft tag

This leaves the following:
* [ ] jonom/focuspoint module support
* [ ] Reflecting changes back to Upload field after the file has been edited
* [ ] Disabling bulk action and checkboxes when editing a file
* [ ] Retain "File selected" counter and Insert Button when editing file in multi-select upload field.

**Comment by maxime-rainville** [comment]
Got the UI fix done. "Reflecting changes back to Upload field after the file has been edited" didn't seem to be worth fixing.

I've rebased and squash the PR. It's ready for re-review.

**Comment by maxime-rainville** [comment]
FYI The build won't pass until to related PRs are merged.

File: tests/php/Forms/FileFormBuilderTest.php [hunk 14]
```diff
@@ -46,6 +46,15 @@ public function setUp()
 
     public function tearDown()
     {
+        // Unpublishing these files is necessary because when this test is run through the recipe-cms phpunit suite,
```
Comment (emteknetnz) [inline tests/php/Forms/FileFormBuilderTest.php:49]: :-/

I get this approach though it's not resolving what seems like a tech-debt situation

Looking at SearchFormTest.php
```
 * Because this manipulates the test database in severe ways, I've renamed the test to force it to run last...
 */
class ZZZSearchFormTest extends FunctionalTest
...
    public static function setUpBeforeClass()
    {
        // HACK Postgres doesn't refresh TSearch indexes when the schema changes after CREATE TABLE
        // MySQL will need a different table type
        static::$tempDB->kill();
        Config::modify();
        FulltextSearchable::enable();
        static::$tempDB->build();
        static::resetDBSchema(true);
        parent::setUpBeforeClass();
    }
```

Seems quite wrong that a unit tests can affect other unit tests.  Would it be possible instead to somehow make what's going on in ZZZSearchFormTest less nasty?

Comment (maxime-rainville) [reply tests/php/Forms/FileFormBuilderTest.php:49]: Not really. `FulltextSearchable::enable()` is not designed to be disabled after it's been enabled. I had a go at getting this working last time I came across this issue and didn't come to an obvious way of fixing it.

And FulltextSearchable is kind of a legacy feature we would probably dump if we had the option right now, so I can't see the logic on spending a lot of time debugging it's current implementation.

My instinct would be to disable this test if it's run as part of the `recipe-cms` suite to avoid these kind of errors going forward.

Comment (emteknetnz) [reply tests/php/Forms/FileFormBuilderTest.php:49]: >> My instinct would be to disable this test if it's run as part of the recipe-cms suite to avoid these kind of errors going forward.

Yup agree there.  I've spend some time trying to replicate this issue locally and I failed to do so.  It's a strange edge-case and probably not worth investing time in fixing properly.

@maxime-rainville do you know of a method to specifically disable this test in a recipe-cms context?

File: client/src/containers/AssetAdmin/Breadcrumb.js [hunk 15]
```diff
@@ -0,0 +1,90 @@
+/* global alert, confirm */
+import React from 'react';
+import PropTypes from 'prop-types';
+import i18n from 'i18n';
+import { Component as PlainBreadcrumb } from 'components/Breadcrumb/Breadcrumb';
```
Comment (emteknetnz) [inline client/src/containers/AssetAdmin/Breadcrumb.js:5]: So this means we would now have two different things called 'Breadcrumb'?

That's going to cause confusion down the road

We should rename the AssetAdmin Breadcrumb to something else like FolderBreadcrumb

File: client/src/containers/Editor/Editor.js (hunk 13 above)
Comment (emteknetnz) [reply client/src/containers/Editor/Editor.js:288]: your github comment says it much better :-) could just copy paste that in

File: client/src/containers/Editor/tests/EditorHeader-story.js (hunk 7 above)
Comment (emteknetnz) [reply client/src/containers/Editor/tests/EditorHeader-story.js:1]: we should now delete this file?

File: code/Forms/FileFormFactory.php [hunk 16]
```diff
@@ -58,15 +49,22 @@ protected function getFormFieldTabs($record, $context = [])
         // All non-admin forms are typically readonly
```
Comment (emteknetnz) [inline code/Forms/FileFormFactory.php:49]: Can remove this comment

File: code/Forms/FileFormFactory.php [hunk 17]
```diff
@@ -58,15 +49,22 @@ protected function getFormFieldTabs($record, $context = [])
...
+            default:
+                $tabs->push($this->getFormFieldDetailsTab($record, $context));
+                $tabs->push($this->getFormFieldSecurityTab($record, $context));
+                $tabs->push($this->getFormFieldUsageTab($record, $context));
+
+                if ($this->config()->get('show_history')) {
+                    $tabs->push($this->getFormFieldHistoryTab($record, $context));
+                }
+
+                $tabs->setReadonly($type === static::TYPE_SELECT);
```
Comment (emteknetnz) [inline code/Forms/FileFormFactory.php:67]: read better then old way i.e.
```
            case static::TYPE_SELECT:
                $tabs->setReadonly(true);
```

File: client/src/containers/AssetAdmin/Breadcrumb.js (hunk 15 above)
Comment (maxime-rainville) [reply client/src/containers/AssetAdmin/Breadcrumb.js:5]: This one is not exposed out of asset-admin, so other modules won't be able to access it. It's only going to be used in the AssetAdmin component.

I see this as having two classes with the same name but under different namespaces.

Comment (emteknetnz) [reply client/src/containers/AssetAdmin/Breadcrumb.js:5]: Yeah it's still confusing though.  I spent about 10 minutes thinking that you'd moved the Breadcrumb Component out of the admin module and put it here :-)

File: client/src/containers/Editor/tests/EditorHeader-story.js (hunk 7 above)
Comment (maxime-rainville) [reply client/src/containers/Editor/tests/EditorHeader-story.js]: Done.

File: client/src/containers/Editor/Editor.js (hunk 13 above)
Comment (maxime-rainville) [reply client/src/containers/Editor/Editor.js]: Done.

File: code/Forms/FileFormFactory.php (hunk 16 above)
Comment (maxime-rainville) [reply code/Forms/FileFormFactory.php]: Done.

File: client/src/containers/AssetAdmin/Breadcrumb.js (hunk 15 above)
Comment (maxime-rainville) [reply client/src/containers/AssetAdmin/Breadcrumb.js]: Done.

File: .travis.yml [hunk 18]
```diff
@@ -59,7 +59,7 @@ before_script:
 # Install composer
   - composer validate
   - if [[ $DB == PGSQL ]]; then composer require --no-update silverstripe/postgresql:2.x-dev --prefer-dist; fi
-  - composer require silverstripe/recipe-testing:^1 silverstripe/recipe-cms 4.x-dev --no-update --prefer-dist
+  - composer require silverstripe/recipe-testing:^1 silverstripe/recipe-cms:4.x-dev silverstripe/frameworktest:dev-master --no-update --prefer-dist
```
Comment (emteknetnz) [inline .travis.yml:62]: As discussed on slack:
- including frameworktest as dev-dependency in composer WILL NOT require frameworktest unless it's the in the project root.
- means it won't do anything on your local if you're doing development from the vendor folder (as most of us are)
- in the case of asset-admin being the root project, which it will in travis, it is functionally the same, though it's clearer being in travis.yml because it shows the intent better 

Seems like the bigger question though is why is frameworktest a requirement for behat tests in the first place?  I've always seen frameworktest as a quick and dirty way to create lots of records for manual testing.  We're including dev-master here because we don't bother versioning frameworktest

Rather then rely on framework test for behat fixture creation, seems like it would be much cleaner to create them the old fashioned way i.e. yml

Comment (maxime-rainville) [reply .travis.yml:62]: There's a precedent for using frameworktests ... it gets installed by the CWP kitchen sink.

I can't just rely on creating fixtures with YML because I need an DataObject with an `has_many` relationship to Image or File. Off the top of my head, I don't think we have a recipe that has that particular set up out of the box

Comment (emteknetnz) [reply .travis.yml:62]: You can probably guess what I'm going to say next :-)

Just create a TestOnly DataObject in asset-admin and use that.  Much cleaner and safer than using frameworktest that has no versioning and could change at any point.

**Comment by emteknetnz** [comment]
Fairly minor UX issue, happy for this to be created as a new issue resolved sometime later on post-merge.

When I edit a page
And I use TinyMCE
And I click the insert an image button
And I upload a new image
And I click the "Details" button in the top right corner
And I change the "Title" of the image
And I click the "Save" button
Then I expect to see the green button to have the text "Insert image"
But actually it has the text "Update image"

**Comment by emteknetnz** [comment]
@maxime-rainville 
One issue we should resolve before merging - the datetime fields on the image are now editable.  They should be read-only

image

On the main branch, they are read-only in the details tab

image

**Comment by clarkepaul** [comment]
> Then I expect to see the green button to have the text "Insert image". But actually it has the text "Update image"

@emteknetnz I can't see the actual experience but from what I recall... as the image is already inserted, you are just updating it. If you were to remove it and re-insert it you get the "insert" button label again.

**Comment by emteknetnz** [comment]
> I can't see the actual experience but from what I recall... as the image is already inserted, you are just updating it. If you were to remove it and re-insert it you get the "insert" button label again.

That's correct for already inserted images.  However you can still get this 'Update image' (or 'Update file' for pdfs) for "uninserted" Images/Files.  It still functions correctly as it will insert the file, it's just it shows the wrong text

**Comment by maxime-rainville** [comment]
I've got a fix for the two fields that should be readonly.

@clarkepaul Would you have a problem if updated the modal so that the button always says "Insert" even if you are updating an image that has previously been inserted? The logic for this part is kind of clunky (which is why it's broken in the first place). I could probably make it work at the cost of adding an extra layer of clunkiness.

If we're OK with the button always saying "Insert", I can actually remove some clunky code.

**Comment by maxime-rainville** [comment]
Just had a Zoom catch up with Paul to discuss how to handle the Insert/Update button issue. The 3 options discussed were:
1. Have the button to always display "Insert" no matter the context.
2. Me spending a couple hours to fix the bug with the button by adding some extra hackish logic on top of the existing hackish logic.
3. Live with the bug.

Paul says he's happy to live with the bug since it's not on the primary path most users will take and it's kind of minor anyway.

**Comment by clarkepaul** [comment]
@maxime-rainville has just walked me through the UX button label bug. I'm okay with leaving the bug as-is for now if we can raise it as a separate issue that would be good.
My reasoning is that for the two most common user paths of:
- inserting an image, the language "insert" makes the most sense
- editing a file when it has already been inserted, the language "update" makes the most sense

From my experience, cms users are more concerned about the presentation of a file initially, editing the file details is a secondary thought (although the main benefits of this issue). Returning without making any edits at all would be subset of people. For the remaining people, I think they would complete the intended action of inserting the file with the button label being "Update" without too many concerns.

Obviously I'd prefer it fixed but I don't see this as a blocker to getting the issue across the line.

**Comment by sachajudd** [comment]
Nice work everyone!! 🎉

**Comment by purplespider** [comment]
This is nice! But may I suggest changing the "**Details**" button to say "**Edit**" instead? To make it more obvious how to edit the fields? 

I'd expect a "Details" button to provide me with more, erm, details, but this button makes the existing fields editable.

Screenshot 2020-06-01 at 16 35 27@2x

**Comment by clarkepaul** [comment]
Much discussion and back and forth on this one. 
- The fields in the current placement view are already editable, you might think you need to click _Edit_ to adjust the current fields in view already.
- With a label _Edit original_ we had comments from users like "What's the original file? why would I want to edit a different file?" The concept of original made it feel like its a copy rather than the same file.

The button takes you to the file _Details_ tab, so having a button with the same name made some sense here. _Edit details_ was also considered but it tested fine without the word edit. This was the terminology which was already being used navigating between the _Placement_ tab and _Details_ tab.

This one doesn't the best experience you could ever want but hopefully it's an improvement from what was there.

**Comment by purplespider** [comment]
Ah, thanks Paul, after reading your comments, I've realised this button appears in two different screens:

---

### 1. Above the Placement tab, when dealing with a WYSIWYG image:

Screenshot 2020-06-02 at 09 42 07@2x

In this context, everything you said makes sense and changing it to just say "Edit" would be unclear.

---

### 2. Above a read-only Details tab, when dealing with an image in an UploadField:

Screenshot 2020-06-02 at 09 43 37m@2x

It is in this context, where the **Details** button is currently confusing. As you are already viewing the Details tab.

---

I presume there is a reason why the fields just aren't immediately editable in the latter view. So, perhaps the best solution is to change the button to say **Edit Details** so that it's action is clear in both contexts?
